torchvision/csrc/io/decoder/gpu/decoder.cpp

```diff
@@ -0,0 +1,428 @@
+#include "decoder.h"
+#include <cassert>
+#include <cmath>
+#include <cstring>
+#include <unordered_map>
+
+static float chroma_height_factor(cudaVideoSurfaceFormat surfaceFormat) {
+  return (surfaceFormat == cudaVideoSurfaceFormat_YUV444 ||
+          surfaceFormat == cudaVideoSurfaceFormat_YUV444_16Bit)
+      ? 1.0
+      : 0.5;
+}
+
+static int chroma_plane_count(cudaVideoSurfaceFormat surfaceFormat) {
+  return (surfaceFormat == cudaVideoSurfaceFormat_YUV444 ||
+          surfaceFormat == cudaVideoSurfaceFormat_YUV444_16Bit)
+      ? 2
+      : 1;
+}
+
+/* Initialise cuContext and videoCodec, create context lock and create parser
+ * object.
+ */
+void Decoder::init(CUcontext context, cudaVideoCodec codec) {
+  cuContext = context;
+  videoCodec = codec;
+  check_for_cuda_errors(cuvidCtxLockCreate(&ctxLock, cuContext), __LINE__);
+
+  CUVIDPARSERPARAMS parserParams = {
+      .CodecType = codec,
+      .ulMaxNumDecodeSurfaces = 1,
+      .ulClockRate = 1000,
+      .ulMaxDisplayDelay = 0u,
+      .pUserData = this,
+      .pfnSequenceCallback = video_sequence_handler,
+      .pfnDecodePicture = picture_decode_handler,
+      .pfnDisplayPicture = picture_display_handler,
+      .pfnGetOperatingPoint = operating_point_handler,
+  };
+  check_for_cuda_errors(
+      cuvidCreateVideoParser(&parser, &parserParams), __LINE__);
+}
+
+/* Destroy parser object and context lock.
+ */
+Decoder::~Decoder() {
+  if (parser) {
+    cuvidDestroyVideoParser(parser);
+  }
+  cuvidCtxLockDestroy(ctxLock);
+}
+
+/* Destroy CUvideodecoder object and free up all the unreturned decoded frames.
+ */
+void Decoder::release() {
+  cuCtxPushCurrent(cuContext);
+  if (decoder) {
+    cuvidDestroyDecoder(decoder);
+  }
+  while (!decodedFrames.empty()) {
+    uint8_t* frame = decodedFrames.front();
+    decodedFrames.pop();
+    cuMemFree((CUdeviceptr)frame);
+  }
+  cuCtxPopCurrent(NULL);
+}
+
+/* Trigger video decoding.
+ */
+void Decoder::decode(const uint8_t* data, unsigned long size) {
+  CUVIDSOURCEDATAPACKET pkt = {
+      .flags = CUVID_PKT_TIMESTAMP,
+      .payload_size = size,
+      .payload = data,
+      .timestamp = 0};
+  if (!data || size == 0) {
+    pkt.flags |= CUVID_PKT_ENDOFSTREAM;
+  }
+  check_for_cuda_errors(cuvidParseVideoData(parser, &pkt), __LINE__);
+  cuvidStream = 0;
+}
+
+/* Fetch a decoded frame and remove it from the queue.
+ */
+uint8_t* Decoder::fetch_frame() {
```

Comment:
Well this function could return nullptr in some cases, when decodedFrames queue is empty(at the start of decoding). This avoids the need to creating unnecessary tensors. Also, I have kept it this way to avoid changing these functions when we support RGB or extend it to other output types.